infrastructure/src/main/java/org/corfudb/infrastructure/log/Compression/Codec.java

```diff
@@ -0,0 +1,46 @@
+package org.corfudb.infrastructure.log.Compression;
+
+import java.nio.ByteBuffer;
+import java.util.function.Supplier;
+
+/**
+ *
+ * An interface that defines generic compress/decompress functionality
+ *
+ * Created by Maithem on 11/6/19.
+ */
+
+public interface Codec {
+
+    /**
+     * Supported codecs
+     */
+    enum Type {
```

Comment:
it seems like a weird singleton, what's the profit? I see it as more complex solution than just using plain old simple singleton

Reply:
Its not. This serves as a mapping between int -> codec type which is later used when the LogData is persisted, so from a binary log prospective 1 -> None, 2 -> LZ4, 3 -> ZSTD.
It also serves as a sanity check, where no two codecs are defined with the same id (since enum prevents that).

Reply:
you just need an enum + a factory method somewhere, but right now your enum is a factory method for himself which makes code much more complex. Also, the lazy initialization doesn't work here,  which brings more complexity but in reality, there is no lazy initialization in this case.

As I said it can be simplified to:
 - enum `Type` contains only values
 - a factory method somewhere: 
   ```
    public Codec getByType(Type codecType) {...}
  ```

Reply:
> Also, the lazy initialization doesn't work here, which brings more complexity but in reality, there is no lazy initialization in this case.

I understand and I wasn't optimizing for that. I can do double checked locking, but it seems like an overkill. If you think that optimizing two very small objects away is crucial, then I'll do it, otherwise I opt for simplicity. 

> lazy initialization doesn't work here, which brings more complexity 

I'm a java neophyte, can you please explain how the lack of lazy initialization introduces complexity? and what type of complexity; Is it performance, efficiency, code organization, correctness,etc?

Reply:
have discussed offline